boards/arm/stm32f429i_disc1/stm32f429i_disc1_defconfig

```diff
@@ -23,6 +23,12 @@ CONFIG_PINMUX=y
 # enable GPIO
 CONFIG_GPIO=y
 
+# enable FMC
+CONFIG_MEMC=y
+
+# enable SDRAM
+CONFIG_MEMC_STM32_SDRAM=y
```

Comment:
There's no need to enable SDRAM by default. Test takes care of that, if there was a sample it should be the sample.

Reply:
So what about enabling it by default for the board? SDRAM is after all board specific and it's always there. Pins connected to SDRAM can't be used for anything else anyway. Is there a logic behind enabling SDRAM at application level and not board level?

Reply:
@gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".

Reply:
Adding this setting to defconfig means that even the hello world sample will compile SDRAM when it doesn't need it. Upstream boards may be used on a variety of samples/tests. Each sample/test needs to take care of enabling the required stuff using overlay files.

Reply:
I think it makes sense to enable support for everything that the board supports at board level. If samples fail it will serveas indication that something is wrong in configuration or that there is a serious conflict. Under normal circumstances, enabling all on board peripherals should not break any samples. (if samples break, it is important to catch these instances asap and fix the conflicts)

Reply:
Ah! I have a solution @gmarull @erwango! The issue is that things get instantiated even when nodes are not present in device tree. The solution is to make sure things *only* getinstantiated and initialized if the corresponding device tree node is there and is status okay. *All* on board peripherals should be enabled on board level in defconfig but device tree nodes should be disabled and then samples and application just sets devicetree nodes to status okay if they are needed. This ensures working configuration for everything supported by the board AND small code size because linker will eliminate dead code automatically for anything that has not been explicitly instantiated even if it has been compiled. Perfect! @MaureenHelm please comment as well.

Reply:
The problem is that Zephyr boards may be used to run a wide variety of samples, and not each sample has the same requirements. In a real application board where you'll likely run a single firmware it makes sense to enable everything at board level.

I think changing current approach would probably deserve an RFC to discuss the implications/side effects, so I think for now it is better to follow the existing approach to have homogeneity in the tree. 

In Zephyr there is a tight relation between DT-Kconfig because of its static approach. So in most cases it does not make sense to have a Kconfig enabled w/o its DT counterpart, and in fact, you'll likely get some compiler warnings/errors. Regarding DT, I think another approach could be to have all board specific stuff defined in the board DT (e.g. SDRAM timings), but still keep the node disabled and enable as needed on samples/tests overlays. Again, this probably needs further discussion as there may be unforeseen implications.

Reply:
> @gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".

This should work. Let's take a basic example:
```
$ west build -b nucleo_f429zi samples/hello_world/
-- west build: generating a build system
[...]
-- west build: building application
[1/150] Preparing syscall dependency handling
[145/150] Linking C executable zephyr/zephyr_prebuilt.elf
[150/150] Linking C executable zephyr/zephyr.elf
```
If I try to rebuild without modification on the board, nothing happens
```
$ west build -b nucleo_f429zi samples/hello_world/
ninja: no work to do.
```
Then I update board .dts to disable one uart node:
```
&usart3 {
	pinctrl-0 = <&usart3_tx_pd8 &usart3_rx_pd9>;
	current-speed = <115200>;
	/* status = "okay"; */
};
```
Launch the build:
```
$ west build -b nucleo_f429zi samples/hello_world/
[0/1] Re-running CMake...
[106/110] Linking C executable zephyr/zephyr_prebuilt.elf
[110/110] Linking C executable zephyr/zephyr.elf
```

So device tree modifications are detected and we should not end up with wrong binary configuration.
If this happens, then there might be issues at some level.


> Ah! I have a solution @gmarull @erwango! The issue is that things get instantiated even when nodes are not present in device tree.

This should not be the case. Instantiation should happen only based on device tree status. Cf upper.

> The solution is to make sure things only getinstantiated and initialized if the corresponding device tree node is there and is status okay. All on board peripherals should be enabled on board level in defconfig but device tree nodes should be disabled and then samples and application just sets devicetree nodes to status okay if they are needed. This ensures working configuration for everything supported by the board AND small code size because linker will eliminate dead code automatically for anything that has not been explicitly instantiated even if it has been compiled. Perfect! @MaureenHelm please comment as well.

First let me explain the strategy we're applying on this board/application configuration split.
Our preoccupation for default configuration for upstream boards is to ensure upstream applications (samples/tests) portability. This is required for:
- demonstrate zephyr modularity and small footprint
- ease CI testing by ensuring maximal portability of these applications. Running one of these applications on a new board should require minimum modification.

To achieve this, responsibilities are split in 2:
- board configuration is done on board side. It is up to the board to provide the default configuration that should be applied (which i2c instance should be used by default). This recommendation is documented here https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#general-recommendations
- Enabling required subsystems to achieve the required functions (CONFIG_I2C=y, CONFIG_LOG=y, ...) is application responsibility

This way, applications maintainers only have to care about the subsystems they need for their application, also running an application on a new board doesn't require any modification on the application.
On their side, boards maintainers should only care about the peripherals the board supports. No dependencies on application.


Current way of working is applied for the +200 boards supported upstream by zephyr and XX samples/tests. This is not perfect and there are still improvements to be done. 
Also, this strategy has limits and you'll happen to find some boards/foo.overlay within samples/ and tests/. This could be required when it's not possible to provide a default board configuration that satisfies all applications. We're trying to minimize these, but I don't think we'll be able to totally getting rid of these. 


Now, let me come back on your proposal. I'm affraid this would not work in our set up.
Let's say you want to be able to run i2c and spi sample on board foo. Default configuration is done on board side (CONFIG_I2C=y, CONFIG_SPI=y) so the board can support the 2 applications.
In order to run `sample/i2c` you'd need to enable _i2c_y node in `sample/i2c/boards/foo.overlay`.
In order to run `sample/spi` you'd need to enable _spi_x node in `sample/spi/boards/foo.overlay`.
At the end you'll get one board overlay per sample (+200 for now, and keeps increasing) which is huge and hardly maintainable while on boards/ we'll get redundant list of CONFIG_FOO activation, which has quite little value.

Reply:
Thanks for this detailed explanation @erwango!

I think that the case it is confusing is for example the following:

Let's take the `samples/drivers/counter/alarm`. In case of STM32, it will use the `#define TIMER DT_LABEL(DT_INST(0, st_stm32_rtc))` device. If one takes for example `stm32h747i_disco_m7` board, sample builds and runs fine regardless of the state of the `rtc` node in DT. In fact, that board doesn't even reference `rtc`, so it should be disabled. This could be prevented if the driver `DEVICE_AND_API_INIT` is guarded with a `status = "okay"` check as Martin suggests, correct me if I'm wrong.

Reply:
> > @gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".
> 
> This should work. Let's take a basic example:
> 

It works there because uart is instantiated by checking ok status "DT_INST_FOREACH_STATUS_OKAY(STM32_UART_INIT)". The memc code just calls a simpler version DEVICE_DEFINE which I think calls init regardless of device tree status. Which is wrong.

Reply:
> > > @gmarull but Kconfig doesn't run if I just change status = "okay" to status = "disabled". So that creates a firmware that misbehaves unless instantiation is also disabled by the DT_ macro that checks for status = "okay".
> > 
> > 
> > This should work. Let's take a basic example:
> 
> It works there because uart is instantiated by checking ok status "DT_INST_FOREACH_STATUS_OKAY(STM32_UART_INIT)". The memc code just calls a simpler version DEVICE_DEFINE which I think calls init regardless of device tree status. Which is wrong.

Multi-instance drivers are guarded today because it turns out they all use the `DT_INST_FOREACH_STATUS_OKAY` macro. I see that most, if not all, single-instance drivers like FMC, counter, flash... don't have an equivalent guard. You're right, I think they should also be guarded after all. @erwango are you aware of any reason why it is not done?

Reply:
@erwango First of all thank you for the really long and detailed explanation. However I need some more clarity here. 

First of all I'm a bit confused with your reply. You say "I'm afraid this would not work in our set up" - which I assumed referred to my reply you quoted - and then you propose that settings like CONFIG_I2C=y, CONFIG_SPI=y should be on board level and defaults should be enabled at board level as well - which is one step further ahead of my point even. I agree with you on that but seem to not be on the same page in this thread. Do we enable I2C and SPI on board level and provide device tree nodes that are enabled by default? Do we provide a generic overlay for sample that enables only aliased devices? Do we even put CONFIG*_STM32* in samples because that would require sample to know which board it is running on? Is that wrong? So which approach is the right one? Is it to enable all board peripherals at board level instead of in sample overlays? 

Secondly, "This should not be the case. Instantiation should happen only based on device tree status. Cf upper.". This is the case unless one uses DT macros that explicitly check for OKAY status. I think this should be made mandatory so that all drivers should use the same pattern as the uart you mentioned. Currently this does not seem to be the case, so when a driver is defined using macro "DEVICE_DEFINE" then there is no check involved so the code runs regardless of state of device tree. Sure, it may not get compiled if Kconfig is not enabling it, but it should not be instantiated and run if it is enabled in Kconfig but device tree node is disabled. Unless DT macro is used to check for status OKAY, it will run anyway which breaks expectation about things actually being disabled when device tree node is disabled.

Reply:
> You're right, I think they should also be guarded after all. @erwango are you aware of any reason why it is not done?

@gmarull No reason. This should be fixed indeed.

> Do we enable I2C and SPI on board level and provide device tree nodes that are enabled by default? Do we provide a generic overlay for sample that enables only aliased devices? Do we even put CONFIG*_STM32* in samples because that would require sample to know which board it is running on? Is that wrong? So which approach is the right one? Is it to enable all board peripherals at board level instead of in sample overlays?

@mkschreder I'm not sure to follow. I've described the approach we're using and some of the arguments behind this approach. If you have alternative to propose, don't hesitate to raise an RFC and provides more details.

Reply:
> @mkschreder I'm not sure to follow. I've described the approach we're using and some of the arguments behind this approach. If you have alternative to propose, don't hesitate to raise an RFC and provides more details.

It makes a bit more sense to keep it this way after seeing how many tests fail if I change it and reading the board porting guide. I'll post RFC if I come up with more concrete changes to be made. Thanks for your clarification.